src/Microsoft.VisualStudio.Composition.Analyzers.CodeFixes/VSMEF004ExportWithoutImportingConstructorCodeFixProvider.cs

```diff
@@ -0,0 +1,228 @@
+    }
+
+    private static async Task<Document> AddImportingConstructorAttributeAsync(
+        Document document,
+        ConstructorDeclarationSyntax constructorDeclaration,
+        MefVersion mefVersion,
+        CancellationToken cancellationToken)
+    {
+        var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+        if (root is null)
+        {
+            return document;
+        }
+
+        // Create the ImportingConstructor attribute
+        var attributeName = mefVersion == MefVersion.V1
+            ? "System.ComponentModel.Composition.ImportingConstructor"
+            : "System.Composition.ImportingConstructor";
+
+        var attribute = SyntaxFactory.Attribute(SyntaxFactory.ParseName(attributeName));
+        var attributeList = SyntaxFactory.AttributeList(SyntaxFactory.SingletonSeparatedList(attribute));
+
+        // Add the attribute to the constructor
+        var newConstructor = constructorDeclaration.WithAttributeLists(
+            constructorDeclaration.AttributeLists.Add(attributeList));
+
+        var newRoot = root.ReplaceNode(constructorDeclaration, newConstructor);
+        return document.WithSyntaxRoot(newRoot);
+    }
+
+    private static async Task<Document> AddParameterlessConstructorAsync(
+        Document document,
+        ClassDeclarationSyntax classDeclaration,
+        CancellationToken cancellationToken)
+    {
+        var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+        if (root is null)
+        {
+            return document;
+        }
+
+        // Create a parameterless constructor
+        var newConstructor = SyntaxFactory.ConstructorDeclaration(classDeclaration.Identifier)
+            .WithModifiers(SyntaxFactory.TokenList(SyntaxFactory.Token(SyntaxKind.PublicKeyword)))
+            .WithParameterList(SyntaxFactory.ParameterList())
+            .WithBody(SyntaxFactory.Block());
+
+        // Find the position to insert the constructor (after existing constructors)
+        var existingConstructors = classDeclaration.Members.OfType<ConstructorDeclarationSyntax>().ToList();
+        int insertIndex = 0;
+
+        if (existingConstructors.Any())
+        {
+            // Insert after the last constructor
+            var lastConstructor = existingConstructors.Last();
+            insertIndex = classDeclaration.Members.IndexOf(lastConstructor) + 1;
+        }
+        else
+        {
+            // Insert at the beginning of the class
+            insertIndex = 0;
+        }
+
+        var newMembers = classDeclaration.Members.Insert(insertIndex, newConstructor);
+        var newClass = classDeclaration.WithMembers(newMembers);
+
+        var newRoot = root.ReplaceNode(classDeclaration, newClass);
+        return document.WithSyntaxRoot(newRoot);
+    }
+
+    private static MefVersion DetermineMefVersion(ClassDeclarationSyntax classDeclaration, SemanticModel semanticModel)
```

Comment:
Remember that a MEF part may have no attributes from either of these namespaces at all, if the only MEF attributes it has are custom ones derived from an ExportAttribute.